core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -0,0 +1,370 @@
+
+  // Scheduler task
+  private var schedulerTask: Option[ScheduledFuture[_]] = None
+
+  // Broker states - Current and Target/Next
+  private var currentState: metadata.BrokerState = _
+  private val pendingHeartbeat = new AtomicBoolean(false)
+
+  // Metrics - Histogram of broker heartbeat request/response time
+  // FIXME: Tags
+  private val heartbeatResponseTime = newHistogram(
+    name = "BrokerHeartbeatResponseTimeMs",
+    biased = true,
+    Map("request" -> "BrokerHeartBeat")
+  )
+  private var _lastSuccessfulHeartbeatTime: Long = 0 // nanoseconds
+
+  /**
+   * Attempts to register the broker
+   * - On success, schedules a periodic heartbeat task
+   * - On failure, throws an exception which can be conditionally retried
+   *
+   * @param listeners - List of endpoints configured for this broker
+   * @param features - List of features supported by this broker
+   * @throws org.apache.kafka.common.errors.AuthenticationException
+   *         - Transport layer authentication errors
+   * @throws org.apache.kafka.common.errors.UnsupportedVersionException
+   *         - Broker version not supported
+   * @throws java.io.IOException
+   *         - Disconnected client/Invalid response from the controller
+   * @throws java.lang.InterruptedException
+   *         - Registration wait was interrupted
+   * @throws java.util.concurrent.TimeoutException
+   *         - Registration wait timed out (max wait: registration.lease.timeout.ms)
+   * @throws org.apache.kafka.common.errors.DuplicateBrokerRegistrationException
+   *         - Duplicate broker ID during registration
+   * @throws org.apache.kafka.common.KafkaException
+   *         - Generic catch all for unknown/unhandled exception(s)
+   *
+   */
+  override def start(listeners: ListenerCollection, features: FeatureCollection): Unit = {
+    // FIXME: Handle broker registration inconsistencies where the controller successfully registers the broker but the
+    //        broker times-out/fails on the RPC. Retrying today, would lead to a DuplicateBrokerRegistrationException
+    currentState = metadata.BrokerState.NOT_RUNNING
+
+    // Initiate broker registration
+    val brokerRegistrationData = new BrokerRegistrationRequestData()
+      .setBrokerId(brokerID)
+      .setFeatures(features)
+      .setListeners(listeners)
+      .setRack(rack)
+    val promise = Promise[Unit]()
+
+    def responseHandler(response: ClientResponse) = {
+      validateResponse(response) match {
+        case None =>
+          // Check for API errors
+          val body = response.responseBody().asInstanceOf[BrokerRegistrationResponse]
+          Errors.forCode(body.data().errorCode()) match {
+            case Errors.DUPLICATE_BROKER_REGISTRATION =>
+              currentState = metadata.BrokerState.NOT_RUNNING
+              promise.tryFailure(Errors.DUPLICATE_BROKER_REGISTRATION.exception())
+            case Errors.NONE =>
+              // Registration success
```

Comment:
The heartbeat needs to send a successful registration event to the broker's metadata listener via something like this:

```
kafkaBroker.brokerMetadataListener.put(OutOfBandRegisterLocalBrokerEvent(brokerEpoch))
```

The broker metadata listener code was just merged.

Reply:
Done